### ISSUE TITLE
optimize compiler flags

### ISSUE THREAD

**Issue by kwgoodman** [kwgoodman/bottleneck#122]

**Comment by kwgoodman** [comment]
`ffast-math` doesn't apply to integers (`f` is for float). So I guess there is no need to remove the `-fno-finite-math-only` for int input arrays.

**Comment by jennolsen84** [comment]
I tried the vectorization flags on the whole project.  I see speedups in nansum and nanmean, etc... but degredation in move_sum.

```
$ cd ../base && make bench && cd ../new/ && make bench
python -c "import bottleneck; bottleneck.bench()"
Bottleneck performance benchmark
    Bottleneck 1.1.0dev; Numpy 1.11.0
    Speed is NumPy time divided by Bottleneck time
    NaN means approx one-third NaNs; float64 and axis=-1 are used

                 no NaN     no NaN      NaN        NaN    
                   (10,)   (1000,1000)   (10,)   (1000,1000)
    nansum         14.4        2.4       14.6        7.4
    nanmean        61.6        3.6       69.6        8.8
    nanstd        130.4        3.1      130.5        7.8
    nanvar        124.5        3.1      125.3        7.9
    nanmin         13.1        0.4       13.2        0.4
    nanmax         13.0        0.4       12.9        0.6
    median         46.4        0.9       58.8        6.6
    nanmedian      42.2        3.1       47.9        9.3
    ss              6.4        1.9        6.6        2.1
    nanargmin      23.4        1.3       24.0        2.7
    nanargmax      24.0        1.3       23.9        2.7
    anynan          5.8        0.8        5.6       80.4
    allnan          5.7      115.8        5.7      110.9
    rankdata       39.1       22.2       38.9       46.2
    nanrankdata    44.4       22.4       41.3       31.3
    partsort        5.6        0.9        5.7        1.2
    argpartsort     3.1        0.8        3.2        0.5
    replace         4.0        1.6        4.1        1.6
    push          114.5       16.9      115.2       18.6
    move_sum      182.6      149.7      187.5      434.1
    move_mean     548.9      132.9      557.3      665.4
    move_std      882.3       69.5      934.1      306.4
    move_var      885.6       99.2      917.6      300.9
    move_min      142.0       22.4      145.0       51.3
    move_max      137.4       21.8      145.9       85.6
    move_argmin   230.3       37.5      244.3      236.4
    move_argmax   233.2       38.4      250.6      294.8
    move_median   364.1       43.9      356.6      137.5
    move_rank     360.2        2.6      402.7       12.5
python -c "import bottleneck; bottleneck.bench()"
Bottleneck performance benchmark
    Bottleneck 1.1.0dev; Numpy 1.11.0
    Speed is NumPy time divided by Bottleneck time
    NaN means approx one-third NaNs; float64 and axis=-1 are used

                 no NaN     no NaN      NaN        NaN    
                   (10,)   (1000,1000)   (10,)   (1000,1000)
    nansum         14.4        4.2       14.5        9.2
    nanmean        67.5        6.1       70.6       10.4
    nanstd        136.6        5.4      137.0       10.2
    nanvar        106.4        5.4      107.4       10.0
    nanmin         12.9        0.4       12.6        0.4
    nanmax         13.3        0.4       13.2        0.6
    median         44.6        0.9       53.8        6.5
    nanmedian      41.6        3.0       46.9        9.4
    ss              5.8        2.1        5.9        2.1
    nanargmin      25.0        1.2       24.2        2.4
    nanargmax      24.8        1.1       24.6        2.4
    anynan          6.1        0.9        6.3       86.9
    allnan          6.3      126.4        6.2      121.0
    rankdata       40.0       23.0       40.0       47.9
    nanrankdata    45.3       23.0       41.1       32.4
    partsort        5.4        0.9        5.4        1.2
    argpartsort     2.9        0.9        3.0        0.7
    replace         4.0        1.5        4.1        1.6
    push          133.1       17.5      132.8       19.5
    move_sum      184.4       94.7      186.9      268.3
    move_mean     544.3      128.6      554.2      278.2
    move_std      921.7       68.6      970.1      266.7
    move_var      901.2       96.4      952.0      267.1
    move_min      136.2       22.2      145.9       50.4
    move_max      138.3       21.6      145.0       78.2
    move_argmin   243.8       38.0      246.8      234.7
    move_argmax   241.6       38.5      244.7      295.6
    move_median   370.9       43.9      363.6      137.8
    move_rank     374.5        2.4      424.3       11.4
```

Here is the diff so far: https://github.com/kwgoodman/bottleneck/compare/master...jennolsen84:compilerflags

**Comment by jennolsen84** [comment]
forgot to mention: top benchmark is from master, and the bottom one is the one with new compiler flags

**Comment by kwgoodman** [comment]
I'm surprised how slow nanmin (relative to np) runs on your computer. I get:
```
In [1]: bn.bench(functions=['nanmin', 'nanmax'])
Bottleneck performance benchmark
    Bottleneck 1.1.0dev; Numpy 1.11.0
    Speed is NumPy time divided by Bottleneck time
    NaN means approx one-third NaNs; float64 and axis=-1 are used

                 no NaN     no NaN      NaN        NaN    
                   (10,)   (1000,1000)   (10,)   (1000,1000)
    nanmin         29.9        1.1       29.6        2.0
    nanmax         22.8        1.1       24.7        2.8
```

How much bigger do the flags make the binaries?

So your results suggest that we only use the flags for reduce

**Comment by jennolsen84** [comment]
File sizes seem to be pretty much the same... except for move.so, which was 843KB with the new flags instead of 851KB.  The others were within 1KB and the file sizes are 150KB, 500KB, 1.2MB.

**Comment by jennolsen84** [comment]
what does your nanmin timing look like?

The following numbers are for the master branch (w/o compiler flag changes)

```
import numpy as np
x = np.arange(10000000, dtype=np.float64)

In [3]: %timeit bn.nanmin(x)
100 loops, best of 3: 19.3 ms per loop

In [4]: %timeit np.nanmin(x)
100 loops, best of 3: 6.37 ms per loop
```

**Comment by kwgoodman** [comment]
```
In [1]: x = np.arange(10000000, dtype=np.float64)
In [2]: timeit bn.nanmin(x)
100 loops, best of 3: 6.43 ms per loop
In [3]: timeit np.nanmin(x)
100 loops, best of 3: 6.6 ms per loop
```

As a separate issue this is interesting:
```
In [15]: x = -x
In [16]: timeit bn.nanmin(x)
100 loops, best of 3: 7.81 ms per loop
In [17]: timeit np.nanmin(x)
100 loops, best of 3: 19.1 ms per loop
```

Seems like it is costly to update the min in numpy.

**Comment by jennolsen84** [comment]
My first guess would be the branch predictor, but it is just a guess.  I should look into the bottleneck's implementation though, seems very odd

```
In [3]:  x = np.arange(10000000, dtype=np.float64)
In [4]: timeit bn.nanmin(x)
10 loops, best of 3: 19.2 ms per loop
In [5]: timeit np.nanmin(x)
100 loops, best of 3: 6.45 ms per loop

In [6]: x = -x
In [7]: timeit np.nanmin(x)
100 loops, best of 3: 9.78 ms per loop
In [8]: timeit bn.nanmin(x)
100 loops, best of 3: 19.2 ms per loop

In [10]: x = np.sin(x)
In [11]: timeit bn.nanmin(x)
100 loops, best of 3: 19.3 ms per loop
In [12]: timeit np.nanmin(x)
100 loops, best of 3: 6.43 ms per loop
```

**Comment by jennolsen84** [comment]
Seems like it is my install :( and not the machine.  I grabbed miniconda and install bottleneck, and found this:

```
In [4]: timeit bn.nanmin(x)
100 loops, best of 3: 6.52 ms per loop
```

However, in my own venv, I am still *much* slower, any idea on where I might start exploring?  I might try passing `-march=native` to gcc, but I don't think that is why the bottleneck master is slower on my machine.  

```
backports.shutil-get-terminal-size==1.0.0
Cython==0.24
decorator==4.0.9
entrypoints==0.2.1
flake8==2.5.4
ipykernel==4.3.1
ipython==4.2.0
ipython-genutils==0.1.0
ipywidgets==5.1.3
Jinja2==2.8
jsonschema==2.5.1
jupyter==1.0.0
jupyter-client==4.2.2
jupyter-console==4.1.1
jupyter-core==4.1.0
lz4==0.8.2
MarkupSafe==0.23
mccabe==0.4.0
mistune==0.7.2
nbconvert==4.2.0
nbformat==4.0.1
nose==1.3.7
notebook==4.2.0
numexpr==2.5.2
numpy==1.11.0
pandas==0.18.1
pep8==1.7.0
pexpect==4.0.1
pickleshare==0.7.2
pkg-resources==0.0.0
ply==3.8
ptyprocess==0.5.1
py==1.4.31
pyflakes==1.0.0
Pygments==2.1.3
pytest==2.9.1
python-dateutil==2.5.3
pytz==2016.4
pyzmq==15.2.0
qtconsole==4.2.1
simplegeneric==0.8.1
six==1.10.0
terminado==0.6
tornado==4.3
traitlets==4.2.1
ujson==1.35
widgetsnbextension==1.2.2
yapf==0.7.1
```

```
$ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/5/lto-wrapper
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 5.3.1-14ubuntu2' --with-bugurl=file:///usr/share/doc/gcc-5/README.Bugs --enable-languages=c,ada,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-5 --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-libmpx --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-5-amd64/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-5-amd64 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-5-amd64 --with-arch-directory=amd64 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 5.3.1 20160413 (Ubuntu 5.3.1-14ubuntu2)
```

I am not using MKL, but if I did, I am guessing it would only help numpy and not bottleneck.
```
np.__config__.show()
openblas_lapack_info:
    define_macros = [('HAVE_CBLAS', None)]
    libraries = ['openblas', 'openblas']
    language = c
    library_dirs = ['/usr/local/lib']
openblas_info:
    define_macros = [('HAVE_CBLAS', None)]
    libraries = ['openblas', 'openblas']
    language = c
    library_dirs = ['/usr/local/lib']
blas_mkl_info:
  NOT AVAILABLE
blas_opt_info:
    define_macros = [('HAVE_CBLAS', None)]
    libraries = ['openblas', 'openblas']
    language = c
    library_dirs = ['/usr/local/lib']
lapack_opt_info:
    define_macros = [('HAVE_CBLAS', None)]
    libraries = ['openblas', 'openblas']
    language = c
    library_dirs = ['/usr/local/lib']
```

**Comment by kwgoodman** [comment]
Did you install bottleneck from source with miniconda? Or the binary, which is bottleneck 1.0.

**Comment by jennolsen84** [comment]
binary

**Comment by jennolsen84** [comment]
I guess I should try 1.0 release from source, and if i get the slowdown, then it's the build.  If it is fast, then it could be the code changes

**Comment by jennolsen84** [comment]
Still slow on v1.0.0 release (tried from git, pypi, and pip installing the release directly)

```
$ pip install https://github.com/kwgoodman/bottleneck/archive/v1.0.0.tar.gz
Collecting https://github.com/kwgoodman/bottleneck/archive/v1.0.0.tar.gz
  Downloading https://github.com/kwgoodman/bottleneck/archive/v1.0.0.tar.gz (366kB)
    100% |████████████████████████████████| 368kB 3.6MB/s 
Requirement already satisfied (use --upgrade to upgrade): numpy in /home/navreet/work/lib/python3.5/site-packages (from Bottleneck===1.0.0)
Installing collected packages: Bottleneck
  Running setup.py install for Bottleneck ... done
Successfully installed Bottleneck-1.0.0
```

```
In [1]: import bottleneck as bn

In [2]: import numpy as np

In [3]: x = np.arange(10000000, dtype=np.float64)

In [4]: timeit bn.nanmin(x)
100 loops, best of 3: 19.3 ms per loop

In [5]: bn.__version__
Out[5]: '1.0.0'
```

**Comment by kwgoodman** [comment]
Are there any other computers around for you to test on?

When I run `make all` here are the compiler flags used:
```
x86_64-linux-gnu-gcc -pthread -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fPIC -I/usr/local/lib/python2.7/dist-packages/numpy/core/include -I/usr/include/python2.7 -c bottleneck/reduce.c -o build/temp.linux-x86_64-2.7/bottleneck/reduce.o
```

Are you using py2.7? 

Odd that only nanmin/nanmax are slow on your system.

**Comment by jennolsen84** [comment]
python3, and the miniconda install was python3 as well.

Here are my relevant compile flags (from master) for reduce.o:  

```
-pthread -DNDEBUG -g -fwrapv -O3 -fPIC
```

Hmm, not sure why mine is using `-O3` vs `-O2`.  I don't see that in setup.py .  According to https://groups.google.com/forum/#!topic/cython-users/Z0jTf5jVOUM, distutils uses whatever python was compiled with.

I modified setup.py to use `-O2` instead of `-O3` and now I see ~6.8ms for nanmin!

**Comment by kwgoodman** [comment]
Nice find.

If I add `-O3` I get a slow nanmin too (18.9ms).
 
I think we should force `-O2` if it doesn't hurt the rest of the benchmark on your computer. I'd like to see benchmark results on more systems. But no one every sends those.

**Comment by kwgoodman** [comment]
`-O2`:
```
In [1]: x = np.arange(10000000, dtype=np.float64)
In [2]: timeit bn.nanmin(x)
100 loops, best of 3: 6.36 ms per loop
In [3]: timeit bn.nanmin(x)
100 loops, best of 3: 6.39 ms per loop
```

`-O1`:
```
In [2]: timeit bn.nanmin(x)
100 loops, best of 3: 6.27 ms per loop
In [3]: timeit bn.nanmin(x)
100 loops, best of 3: 6.25 ms per loop
```

`-O0`:
```
In [2]: timeit bn.nanmin(x)
10 loops, best of 3: 28.5 ms per loop
In [3]: timeit bn.nanmin(x)
10 loops, best of 3: 28.4 ms per loop
```

If we do any more playing with compiler options then we should probably write a comprehensive benchmarking tool. Maybe each function could be put through a battery of tests. And then the benchmark for the function would be the the time to run the battery of tests. Running a (1000, 1000) array with axes 0, 1, None for each dtype would be a good start.

**Comment by jennolsen84** [comment]
I was thinking of the same thing.   I thought about possibly looping through a collection of compiler flags for each module, and selecting the best flags, etc.  But, then I realized that perhaps profile guided optimization would be best.

This would lead to longer build times, but close to optimal performance for those who want it.  For those who want a faster build and decent performance, one could just stick with `-O2`.

One thing we'll also need is good data sets, and a distribution of them representing real world (maybe mostly random, and one strictly increasing and another strictly decreasing?).  This way the profile doesn't memorize the data pattern.

**Comment by kwgoodman** [comment]
We could compile each function separately for optimum performance. Perhaps different reduction functions perform better with different flags. (It would be good to know that that is true before proceeding.)

The inner workings of bottleneck (eg the reducer function) and the top level functions (eg nansum) could be compiled only once with some default flags we pick. Then we'd only have to optimize the small inner functions. In a proposed rewrite (#123) those inner functions are small and therefore, I imagine, quick to compile and link.

Or is that crazy to optimize each function separately?

**Comment by kwgoodman** [comment]
The new proposal uses inline functions. Not sure if splitting those off into separate files will preserve the inlining. Probably better to start by optimizing at the module level. And maybe only optimize reduce and move.

**Comment by kwgoodman** [comment]
Timing long running functions (and short) can be noisy as in this demo:
```
import numpy as np
from bottleneck.tests.functions import reduce_functions
from bottleneck.benchmark.autotimeit import autotimeit


def timeit(array_list, axes=(0, 1, None)):
    for arr in array_list:
        for func in reduce_functions():
            for axis in axes:
                func(arr, axis=axis)


if __name__ == '__main__':
    rs = np.random.RandomState(1)
    a = rs.rand(1000, 1000)
    dtypes = [np.int32, np.int64, np.float32, np.float64]
    array_list = [a.astype(t) for t in dtypes]
    seconds = autotimeit(lambda: timeit(array_list), repeat=10, mintime=0.3)
    print seconds
```

**Comment by kwgoodman** [comment]
I forced the `-O2` flag: https://github.com/kwgoodman/bottleneck/commit/b3e618257d44cb9bc1326083cbd50259b08869ec

And I took care of the NPY_CORDER thing: https://github.com/kwgoodman/bottleneck/commit/d0828eff3b367f452fd39556691a07dae55c1d40

**Comment by jennolsen84** [comment]
I've been looking at PGO with python.  I couldn't get python to generate the profile information when bottleneck was a python library.  I compiled both python and bottleneck to generate the profile information, and while I was able to see profile information generated for python's built in module, I did not see it for bottleneck.

Perhaps it would be easier to actually create another C/C++ program that uses the C version of bottleneck and we can use that to generate the profiles.  

The other option to try is https://docs.python.org/3/extending/embedding.html
